application/app.py

```diff
@@ -1,44 +1,230 @@
+        token_request_base_url = "https://github.com/login/oauth/access_token"
+        token_request_url, token_request_headers, token_request_body = client.prepare_token_request(
+            token_request_base_url,
+            authorization_response=authorization_response_url,
+            code=authorization_response_code,
+            client_secret = client_secret#,
+            # state = authorization_response_state
+        )
+
+        # Get an access token.
+        token_request_response = requests.post(
+            token_request_url,
+            headers = token_request_headers,
+            data = token_request_body
+        )
+        client.parse_request_body_response(token_request_response.text)
+
+        # Get user details.
+        user_request_base_uri = "https://api.github.com/user"
+        user_request_uri, user_request_headers, user_request_body = client.add_token(
+            user_request_base_uri
+        )
+        user_info = requests.get(
+            user_request_uri,
+            headers = user_request_headers,
+            data = user_request_body
+        ).json()
+
+        # Create user object and log user in.
+        user_id = user_info['id']
+        user = User(user_id, "reader")
+        users = mongo_client.users
+        # Assign the first user to be an admin.
+        if len(users) == 0:
+            user.role = "admin"
+        mongo_client.add_user(user_id, user.role)
+        login_user(user)
+
+        # Return to the main page.
+        return redirect('/')
+
+    def admins_only(view_function):
+        @wraps(view_function)
+        def wrapper(*args, **kwargs):
+            if app.config['LOGIN_DISABLED'] == True:
+                return view_function(*args, **kwargs)
+            else:
+                if current_user.is_anonymous:
+                    return redirect('/login')
+                elif current_user.role == 'admin':
+                    return view_function(*args, **kwargs)
+                else:
+                    return redirect('/forbidden')
+        return wrapper
+
+    def admins_and_writers_only(view_function):
+        @wraps(view_function)
+        def wrapper(*args, **kwargs):
+            if app.config['LOGIN_DISABLED'] == True:
+                return view_function(*args, **kwargs)
+            else:
+                if current_user.is_anonymous:
+                    return redirect('/login')
+                elif current_user.role in ('admin', 'writer'):
+                    return view_function(*args, **kwargs)
+                else:
+                    return redirect('/forbidden')
+        return wrapper
+
+    # Login/logout/authorisation screens.
+
+    @app.route('/login')
+    def login():
+        return render_template('login.html')
+
+    @app.route('/logout')
+    def logout():
+        logout_user()
+        return redirect('/login')
+
+    @app.route('/forbidden')
+    def forbidden():
+        return render_template('forbidden.html')
```

Comment:
Worth noting that this will show the relevant page, but will show a 200 success status code currently. It would be better to ensure the status code matches the response - see [the flask docs](https://flask.palletsprojects.com/en/1.1.x/patterns/errorpages/) for how we might do that